### PR TITLE
Kernel 5.11 support

### LINKED ISSUE #276: When snapshot sda2, does sda2 need to change its mrf if it's already changed?
Hi

Disk conf: `sda1`, `sda2`
Steps:
1. First snapshot `sda1`;
2. In `__tracer_transition_tracing` , `sda1->sda->queue->make_request` is replaced with `tracing_mrf`, and so is `sda2->sda->queue->make_request`;
3. Snapshot the `sda2`,  here's the point: **We still run `__tracer_transition_tracing` for sda2**, which will again:
(1) Freeze the `sda`;
(2) Change `sda2->sda->queue->make_request` to `tracing_mrf`. but it already is in step 2.

Could we do some precheck to prevent this?

Best Regards,
Nick

### PR DESCRIPTION
**Pull request by dakotarwilliams**
Built on work from #280, this implementation uses ftrace rather than gendisk swapping to intercept disk I/O. This approach still avoids using unexported functions and adding more symbol lookups at compile-time.

The ftrace approach also avoids the issue we've seen with #276, since there's no function pointer swapping going on. This MR still includes the fix for earlier kernel versions.

This approach could be extended to mount/unmount events thereby re-enabling unverified and dormant snapshot states, however further investigation is required.

Obsoletes #280
Fixes #276

### CODE REVIEW

**Review by Conan-Kudo — APPROVED** [review]

**Comment by ibambotdatto** [comment]
Can't hit the approve button anymore, but aside from the above nits, everything LGTM as far as I could tell.

File: src/callback_refs.c [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                // kzalloc guarantees mtd->mtd_count = { 0 };
+                hash_add(mrf_tracking_map, &mtd->node, (unsigned long)disk);
+        }
+
+        atomic_inc(&mtd->mtd_count);
+
+        return 0;
+}
+
+const BIO_REQUEST_CALLBACK_FN* mrf_put(struct block_device *disk)
```
[the remark this answers is not in the document]
Comment (dakotarwilliams) [reply src/callback_refs.c:56]: get/put is common terminology for reference counted resources. For example: https://elixir.bootlin.com/linux/latest/source/include/linux/percpu-refcount.h
We use similar terminology when accessing our tracing_params struct as well.

File: src/configure-tests/feature-tests/blk_mq_submit_bio.c [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+/*
+ * Copyright (C) 2022 Datto Inc.
+ */
+
+#include "includes.h"
+
+static inline void dummy(void){
```
[the remark this answers is not in the document]
Comment (dakotarwilliams) [reply src/configure-tests/feature-tests/blk_mq_submit_bio.c:9]: Good catch on these, made #289

File: src/callback_refs.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibambotdatto) [reply src/callback_refs.c:56]: Gotcha, learned something new. Thank you!

**Pull request merged**